service/src/main/kotlin/fi/espoo/evaka/shared/config/SpringMvcConfig.kt

```diff
@@ -0,0 +1,74 @@
+// SPDX-FileCopyrightText: 2017-2020 City of Espoo
+//
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+package fi.espoo.evaka.shared.config
+
+import fi.espoo.evaka.identity.ExternalId
+import fi.espoo.evaka.shared.auth.AuthenticatedUser
+import fi.espoo.evaka.shared.auth.getAuthenticatedUser
+import fi.espoo.evaka.shared.db.Database
+import fi.espoo.evaka.shared.domain.Unauthorized
+import fi.espoo.evaka.shared.utils.asArgumentResolver
+import fi.espoo.evaka.shared.utils.convertFromString
+import fi.espoo.evaka.shared.utils.runAfterCompletion
+import org.jdbi.v3.core.Jdbi
+import org.springframework.context.annotation.Configuration
+import org.springframework.core.MethodParameter
+import org.springframework.format.FormatterRegistry
+import org.springframework.web.context.request.NativeWebRequest
+import org.springframework.web.context.request.WebRequest
+import org.springframework.web.context.request.WebRequest.SCOPE_REQUEST
+import org.springframework.web.method.support.HandlerMethodArgumentResolver
+import org.springframework.web.servlet.config.annotation.InterceptorRegistry
+import org.springframework.web.servlet.config.annotation.WebMvcConfigurer
+import javax.servlet.http.HttpServletRequest
+
+/**
+ * Adds support for using the following types as REST function parameters:
+ *
+ * - `Database`: a request-scoped database instance
+ * - `Database.Connection`: a request-scoped database connection, closed automatically at request completion (regardless of success or failure)
+ * - `AuthenticatedUser`: user performing the request
+ * - `ExternalId`: an external id, is automatically parsed from a string value (e.g. path variable / query parameter, depending on annotations)
+ */
+@Configuration
+class SpringMvcConfig(private val jdbi: Jdbi) : WebMvcConfigurer {
+    override fun addArgumentResolvers(resolvers: MutableList<HandlerMethodArgumentResolver>) {
+        resolvers.add(asArgumentResolver(::resolveAuthenticatedUser))
+        resolvers.add(asArgumentResolver { _, webRequest -> webRequest.getDatabaseInstance() })
+        resolvers.add(asArgumentResolver { _, webRequest -> webRequest.getOrOpenConnection() })
```

Comment:
I'm not sure what the semantics are exactly, but my guess is it depends somehow on the argument order in the REST function (e.g. `fun foo(db: Database.Connection, user: AuthenticatedUser)` would first open a connection, while `fun foo(user: AuthenticatedUser, db: Database.Connection)` would first resolve the user.

However, it shouldn't matter in practice because if there's no user information available, it should've been caught much earlier by a HTTP filter, so we return 401 very early and don't even enter Spring MVC processing where argument resolvers are used. In other words, `resolveAuthenticatedUser` should in practice never throw Forbidden unless there's a configuration error somewhere (e.g. if a `/public/...` path route has a `user: AuthenticatedUser` argument)